### PR TITLE
[WIP] model struct tag validator

### PR DESCRIPTION
**Pull request by PetarAtanasov**
Created a branch for this task, free up my master branch.

@markbates comments
Overall, I like the idea, but there are some issues that need to be addressed first.

One of the biggest areas I am interested in seeing changed here is to make it so this can be used programmatically, as well as a CLI tool.

Ideally, I would be able to do something like this:

```
v := validate.New("github.com/foo/models")
// add another processor for a tag:
v.AddProcessor("db", func(m Model, t Tag) (ValidationError, error) {})
v.Run() // returns (map[string][]ValidationError, error)
or
v.Run("User", "Widget") // to validate specific models.
```
Just an idea for an API.

### CODE REVIEW

**Comment by PetarAtanasov** [comment]
@markbates I've implemented those changes and are ready to be reviewed.

File: soda/cmd/validate/model.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+	for structName, fields := range m.tags {
+		for _, t := range fields {
+			duplicateErrors := checkForDuplicates(t, fieldsCache)
+
+			if len(duplicateErrors) > 0 {
+				errorss[structName] = append(errorss[structName], duplicateErrors...)
+			}
+
+			for _, processorTag := range m.processors {
```
Comment (PetarAtanasov) [inline soda/cmd/validate/model.go]: Need to filter the processors by the tag their were added to.

File: soda/cmd/validate/model.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+func (m *model) AddDefaultProcessors(tags ...string) {
+
+	if len(tags) == 0 {
+		tags = []string{AllTags}
+	}
+
+	for _, tagStr := range tags {
+		m.processors[tagStr] = append(m.processors[tagStr], func(tag *Tag) ([]ValidationError, error) {
+			regexpr := []*regexp.Regexp{
```
Comment (stanislas-m) [inline soda/cmd/validate/model.go]: Can you move this outside the loop, if it's constant?

File: soda/cmd/validate/model.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+const AllTags = "*"
+
+type model struct {
+	packages map[string]*ast.Package
+	tags     map[string][]*Tag
+	processors map[string][]func(tag *Tag) ([]ValidationError, error)
+	path string
+}
+
+type ValidationError struct {
```
Comment (stanislas-m) [inline soda/cmd/validate/model.go]: Please add a comment doc on each public struct / function.

File: soda/cmd/validate/model.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	structName     string
+	duplicate      bool
+	fieldName      string
+}
+
+func (e *ValidationError) Error() string {
+	return fmt.Sprintf("Invalid symbols '%v' contained in %v.%v.%v", e.invalidSymbols, e.structName, e.fieldName, e.field)
+}
+
+func (m *model) AddDefaultProcessors(tags ...string) {
```
Comment (stanislas-m) [inline soda/cmd/validate/model.go]: Please add a comment doc on each public struct / function.

File: soda/cmd/validate/model.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	fieldsCache[cacheKey] = true
+
+	return errorss
+}
+
+func (m *model) setPath(path string)  {
+	m.path = path
+}
+
+func NewValidator(path string) model {
```
Comment (stanislas-m) [inline soda/cmd/validate/model.go]: Please add a comment doc on each public struct / function.

File: soda/cmd/validate/model.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	m := model{}
+	m.setPath(path)
+	m.processors =  map[string][]func(tag *Tag) ([]ValidationError, error){}
+
+	return m
+}
+
+func (m *model) Run(models ...string)  (map[string][]ValidationError, error) {
```
Comment (stanislas-m) [inline soda/cmd/validate/model.go]: Please add a comment doc on each public struct / function.

File: soda/cmd/validate/model.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+					errorss[structName] = append(errorss[structName], errs...)
+				}
+			}
+		}
+	}
+
+	return errorss, nil
+}
+
+func (m *model) AddProcessor(tag string, processor func(t *Tag) ([]ValidationError, error)) {
```
Comment (stanislas-m) [inline soda/cmd/validate/model.go]: Please add a comment doc on each public struct / function.

File: soda/cmd/validate/tag.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/pkg/errors"
+)
+
+type Tag struct {
+	name string
+	value string
+	structName string
+}
+
+func (t *Tag) getName() string  {
```
Comment (stanislas-m) [inline soda/cmd/validate/tag.go]: Do you really need this getter?

File: soda/cmd/validate/tag.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+	return t.name
+}
+
+func (t *Tag) getValue() string  {
```
Comment (stanislas-m) [inline soda/cmd/validate/tag.go]: Do you really need this getter?

File: soda/cmd/validate/tag.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+	return t.value
+}
+
+func (t *Tag) getStructName() string  {
```
Comment (stanislas-m) [inline soda/cmd/validate/tag.go]: Do you really need this getter?

File: soda/cmd/validate/tag.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+	return t.structName
+}
+
+func getPackages(folder string, models ...string) map[string]*ast.Package {
+	path := os.Getenv("GOPATH")
```
Comment (stanislas-m) [inline soda/cmd/validate/tag.go]: You'll probably want to use `envy.GoPaths` instead, and remember you can have more than one `
GOPATH`.

File: soda/cmd/validate/tag.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+		modelMap[k] = true
+	}
+
+	pkgs, err := parser.ParseDir(fset, path, func(f os.FileInfo) bool {
+		isNotTest := !strings.HasSuffix(f.Name(), "_test.go")
+
+		if len(modelMap) > 0 {
+			_, exists := modelMap[strings.ToLower(f.Name())]
+			//I want XOR, != is not natural
```
Comment (stanislas-m) [inline soda/cmd/validate/tag.go]: Do the truth table for both operators on paper. ;)

**Comment by PetarAtanasov** [comment]
@stanislas-m changes made, it's ready for a review.

**Comment by stanislas-m** [comment]
> Why aren’t we using gobuffalo/validate. Are we reinventing a library we already maintain? We should probably use that.

`gobuffalo/validate` is about data validation, here it's about code/struct validation. :)

**Comment by markbates** [comment]
Data is data though. It’s a generic validation system. There’s no difference saying this field is empty and a tag being misformed. 

If we use the validate package then it also offers up interoperability with everything else that is currently using it. Mean, if people wanted they could add these validators to their own validators in development mode. 

We have a generic validation package, we should absolutely be using it, unless someone can give me a good reason why not.

**Comment by stanislas-m** [comment]
@markbates That's a valid point. :)

**Comment by markbates** [comment]
I have those occasionally. :)

**Comment by markbates** [comment]
To answer the questions asked in slack:

gobuffalo/validate is already a dependency of pop, so that shouldn’t be an issue. 

Write a validator for each think you want to validate. ie one for the has_many tag, one for db, etc…

The ClI should run all the validators, while as developer I should be able to combine those validators in anyway I would like.

**Comment by PetarAtanasov** [comment]
@markbates 

>  Write a validator for each think you want to validate. ie one for the has_many tag, one for db, etc…

This is a general tag validator, new tags for validation could be added from the user.
I don't see how this fits into the structure.

**Comment by markbates** [comment]
Pop has a lot of tags, db, has_many, etc… if we’re not validating those tags, then I don’t understand the point of this.

**Comment by markbates** [comment]
To clarify there should be a DBTagValidator, a HasManyTagValidator, etc…

**Comment by PetarAtanasov** [comment]
@markbates 
It does validate any specified tag against the user passed lambda.
I am making it work with the gobuffalo/validate validation now.

I will make the lambda processors passed by the users to have exactly the same definition as the gobuffalo/validate ones , so there won't be confusion.

Then i will add those specific tag validators to be provided as defaults.

**Comment by PetarAtanasov** [comment]
@markbates Currently it supports by default, validation of all tags present in a model.
`const AllTags = "*"`

```
func (v *Validator) AddDefaultProcessors(tags ...string) {

	if len(tags) == 0 {
		tags = []string{AllTags}
	}
```

If no tags are passed, all tags are validated.

**Pull request closed without merging**